admin/deployments.go

```diff
@@ -478,107 +477,42 @@ type provisionRuntimeOptions struct {
-	}
-	if ok && pr.Provisioner != opts.Provisioner {
-		return nil, fmt.Errorf("provisioner: cannot change provisioner from %q to %q for deployment %q", pr.Provisioner, opts.Provisioner, opts.DeploymentID)
-	}
-
-	// If we didn't find an existing DB entry, create one
-	if !ok {
-		pr, err = s.DB.InsertProvisionerResource(ctx, &database.InsertProvisionerResourceOptions{
-			ID:            uuid.New().String(),
-			DeploymentID:  opts.DeploymentID,
-			Type:          string(provisioner.ResourceTypeRuntime),
-			Name:          "", // Not giving runtime resources a name since there should only be one per deployment.
-			Status:        database.ProvisionerResourceStatusPending,
-			StatusMessage: "Provisioning...",
-			Provisioner:   opts.Provisioner,
-			Args:          args.AsMap(),
-			State:         nil, // Will be populated after provisioning
-			Config:        nil, // Will be populated after provisioning
-		})
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	// Provision the runtime
-	r := &provisioner.Resource{
-		ID:     pr.ID,
-		Type:   provisioner.ResourceTypeRuntime,
-		State:  pr.State,  // Empty if inserting
-		Config: pr.Config, // Empty if inserting
-	}
-	r, err = p.Provision(ctx, r, &provisioner.ResourceOptions{
-		Args:        args.AsMap(),
-		Annotations: opts.Annotations,
-		RillVersion: s.resolveRillVersion(),
-	})
-	if err != nil {
-		ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
-		defer cancel()
-		_, _ = s.DB.UpdateProvisionerResource(ctx, pr.ID, &database.UpdateProvisionerResourceOptions{
-			Status:        database.ProvisionerResourceStatusError,
-			StatusMessage: fmt.Sprintf("Failed provisioning runtime: %v", err),
-			Args:          pr.Args,
-			State:         pr.State,
-			Config:        pr.Config,
-		})
-		return nil, err
-	}
-
-	// Update the provisioner resource
-	pr, err = s.DB.UpdateProvisionerResource(ctx, pr.ID, &database.UpdateProvisionerResourceOptions{
-		Status:        database.ProvisionerResourceStatusOK,
-		StatusMessage: "",
-		Args:          args.AsMap(),
-		State:         r.State,
-		Config:        r.Config,
+	// Call into the generic provision function
+	pr, err := s.Provision(ctx, &ProvisionOptions{
+		DeploymentID: opts.DeploymentID,
+		Type:         provisioner.ResourceTypeRuntime,
+		Name:         "", // Not giving runtime resources a name since there should only be one per deployment.
+		Provisioner:  opts.Provisioner,
+		Args:         args.AsMap(),
+		Annotations:  opts.Annotations,
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	// Await the runtime to be ready
-	err = p.AwaitReady(ctx, r)
-	if err != nil {
-		return nil, err
-	}
-
 	return pr, nil
 }
 
 func (s *Service) findProvisionedRuntimeResource(ctx context.Context, deploymentID string) (*database.ProvisionerResource, bool, error) {
-	prs, err := s.DB.FindProvisionerResourcesForDeployment(ctx, deploymentID)
+	pr, err := s.DB.FindProvisionerResourceByName(ctx, deploymentID, string(provisioner.ResourceTypeRuntime), "")
```

Comment:
Okay, I changed the query name to `...ByTypeAndName`. 

The runtime resource is not currently named because there is only supposed to be one resource of that type for each deployment. This is different than for ClickHouse, where it matches to the connector name in the runtime and you could in theory have multiple ClickHouses for one deployment (e.g. if you define `clickhouse-a.yaml` and `clickhouse-b.yaml`).